File: scripts/mapgeom.py
"""Reusable map geometry, walk-cost, and zone-name-resolution helpers.

This module depends only on the Python standard library so both travel derivation and map-pack
conversion can import it without creating a repository-internal import cycle.

Most helpers operate only on records supplied by the caller. ``cost_points`` deliberately
memoises sampled points into those zone records under the private ``_cpts`` or ``_cpts_t`` key;
callers must not serialize a record after passing it here without removing private keys.
"""
import math
import re


COST_SAMPLE = 200         # outline points per zone for the closest-approach cost scan
UNITS_PER_COST = 250.0


def tpoint(z, x, y):
    """Apply a zone's render-time affine, exactly as the viewer's tPoint() does."""
    xf = z.get("xf") or {}
    s = xf.get("s", 1)
    s = 1 if s is None else s
    rot = xf.get("rot", 0)
    dx, dy = (x - z["cx"]) * s, (y - z["cy"]) * s
    c, si = math.cos(rot), math.sin(rot)
    return (z["cx"] + dx * c - dy * si + xf.get("tx", 0),
            z["cy"] + dx * si + dy * c + xf.get("ty", 0))


def tinv(z, X, Y):
    """Inverse of tpoint: a world point back into the zone's LOCAL frame."""
    xf = z.get("xf") or {}
    s = xf.get("s", 1)
    s = 1 if s is None else s
    rot = xf.get("rot", 0)
    dx, dy = X - z["cx"] - xf.get("tx", 0), Y - z["cy"] - xf.get("ty", 0)
    c, si = math.cos(-rot), math.sin(-rot)
    return (z["cx"] + (dx * c - dy * si) / s, z["cy"] + (dx * si + dy * c) / s)


_SPELL = [("forrest", "forest"), ("excile", "exile"), ("cablis", "cabilis"),
          ("toxullia", "toxxulia"), ("feerott", "feerrott"), ("aquaduct", "aqueduct"),
          ("northern", "north"), ("southern", "south"), ("eastern", "east"),
          ("western", "west")]
ZALIAS = {"butcherblock": "butcherblock mountains", "kerra ridge": "kerra isle",
          "castle of mistmoore": "castle mistmoore", "city of guk": "upper guk",
          "erudin city": "erudin", "erudin docks": "erudin",
          "north ro": "north desert of ro", "south ro": "south desert of ro",
          "permafrost keep": "permafrost caverns", "temple of cazic-thule": "cazic-thule",
          "skyshrine lower level": "skyshrine", "ruins of old guk": "lower guk",
          "ruins of old paineel": "warrens", "ruins of sebilis": "old sebilis",
          "city of thurgadin": "thurgadin", "qeynos aqueduct system": "qeynos catacombs",
          "liberated citadel of runnyeye": "runnyeye citadel",
          "valley of king xorbb": "gorge of king xorbb"}
LINK_OVERRIDE = {"kithicor|to_The_Commonlands": "commons",
                 "befallen|to_The_Commonlands": "commons"}


def znorm(s):
    s = s.lower().replace("`", "'").replace("_", " ")
    s = re.sub(r"^\s*(to|from)\s+", "", s)
    s = re.sub(r"\(.*?\)", "", s)
    s = re.sub(r":.*$", "", s)
    s = re.sub(r"\bone[- ]way\b", "", s).replace("&", " ").replace(" - ", " ")
    for a, b in _SPELL:
        s = s.replace(a, b)
    s = s.replace("plains of karana", "karana")
    s = re.sub(r"^(the|clan)\s+", "", s)
    return re.sub(r"\s+", " ", s).strip(" -")


def zidx_from(entries):
    """Build a first-wins name index from ``(continent, zone key, name)`` entries."""
    idx = {}

    def put(k, v):
        k = k.strip()
        if k and k not in idx:
            idx[k] = v

    for cont, zk, nm in entries:
        nl = nm.lower()
        put(znorm(nm), (cont, zk))
        mm = re.search(r"\((north|south|east|west)\)", nl)
        if mm:
            b = znorm(re.sub(r"\s*\((north|south|east|west)\)", "", nl))
            put(mm.group(1) + " " + b, (cont, zk))
            put(b + " " + mm.group(1), (cont, zk))
        m2 = re.search(r"neriak \((.*)\)", nl)
        if m2:
            put("neriak " + m2.group(1), (cont, zk))
    return idx


def resolve_zone(idx, lab):
    n = znorm(lab)
    if n in idx:
        return idx[n]
    a = ZALIAS.get(n)
    return idx.get(a) if a else None


def transition_targets(idx, zk, full):
    """Zone keys a detail label is a doorway to. Mirrors the viewer's transitionTargets()."""
    if not re.match(r"^(to|from)_", full, re.I):
        return []
    ov = LINK_OVERRIDE.get(zk + "|" + full)
    if ov:
        return [ov]
    amp = full.find("&")
    if amp < 0:
        t = resolve_zone(idx, full)
        return [t[1]] if t else []
    out = []
    for part in (full[:amp], full[amp + 1:]):
        t = resolve_zone(idx, part)
        if t:
            out.append(t[1])
    return out


def detail_offset(z, det):
    """Detail frame -> continent frame. A pure translation, recovered from one segment pair.

    Confirmed against a second, well-separated segment before it is trusted: one pair is enough
    to RECOVER a translation and not enough to know there is one, and a re-import that reordered
    or dropped a segment would give a plausible offset that silently misplaced every exit in the
    zone. Mirrors the viewer's detailOffset(), including giving up rather than guessing.
    """
    if not z["segs"] or not det["segs"]:
        return None
    ox = z["segs"][0][0] - det["segs"][0][0]
    oy = z["segs"][0][1] - det["segs"][0][1]
    m = min(len(z["segs"]), len(det["segs"])) // 2
    a, b = z["segs"][m], det["segs"][m]
    if abs((a[0] - b[0]) - ox) > 1.5 or abs((a[1] - b[1]) - oy) > 1.5:
        return None
    return ox, oy


def exit_points_from(zk, zone_record, composed_detail, idx):
    """Return ``{(from_key, to_key): (x, y)}`` from one zone's composed detail.

    An unrecoverable detail offset returns an empty mapping. If more than one label names the same
    target, ``setdefault`` preserves the first label in the caller-supplied detail order.
    """
    off = detail_offset(zone_record, composed_detail)
    if not off:
        return {}
    out = {}
    for lab in composed_detail["labels"]:
        for tgt in transition_targets(idx, zk, lab[4]):
            out.setdefault((zk, tgt), (lab[0] + off[0], lab[1] + off[1]))
    return out


def nearest_outline_point(z, px, py, transformed):
    """The point of z's outline nearest (px,py), in the caller's working frame.

    Scans EVERY segment endpoint rather than cost_points()' 200-point sample. That sample exists
    for the O(n*m) closest-approach scan; this is O(n), so thinning it buys nothing and only costs
    accuracy - and it runs at most twice per one-sided edge, of which there are two.
    """
    best, bp = float("inf"), (z["cx"], z["cy"])
    for s in z["segs"]:
        for p in ((s[0], s[1]), (s[2], s[3])):
            q = tpoint(z, p[0], p[1]) if transformed else p
            d = (q[0] - px) ** 2 + (q[1] - py) ** 2
            if d < best:
                best, bp = d, q
    return bp


def cost_points(z, transformed):
    """Outline points for the cost measurement, thinned for an O(n*m) closest-pair scan.

    Separate from ``_pts`` because that one is always transformed and sampled for weld detection.
    A few units of imprecision is irrelevant after dividing by UNITS_PER_COST. This function
    memoises its result into the caller's record under ``_cpts`` or ``_cpts_t``.
    """
    key = "_cpts_t" if transformed else "_cpts"
    if key not in z:
        pts = []
        for s in z["segs"]:
            pts.append((s[0], s[1]))
            pts.append((s[2], s[3]))
        if len(pts) > COST_SAMPLE:
            step = len(pts) / COST_SAMPLE
            pts = [pts[int(i * step)] for i in range(COST_SAMPLE)]
        z[key] = [tpoint(z, p[0], p[1]) for p in pts] if transformed else pts
    return z[key]


def cost_between(zones, k1, k2, transformed, exits=None):
    """Default walk cost: centroid -> the doorway each zone names -> centroid, / UNITS_PER_COST.

    NOT centroid separation. The continental stitch does not place zones at true relative
    distance - Innothule Swamp and South Desert of Ro are drawn 13 800 units apart with
    nothing between them - so centroid separation bills the traveller for empty canvas. A
    connector between two zones means they are directly connected, whatever the drawing puts
    between them, so the void must not be priced.

    The doorway comes from the zone's OWN detail map: its `to_X`/`from_X` label marks where that
    exit physically is, resolved through the viewer's name index and placed on the continent map
    by detail_offset(). On current data 72 of 74 edges name it from both sides and none from
    neither.

    Geometric closest approach is the FALLBACK, because it is only a valid proxy for a zone line
    where the outlines actually touch. Where a stitch void separates them it returns whichever
    flanks happen to face each other - a fact about relative box placement, not about the game.
    It put the South Ro crossing on Innothule's east flank beside the Guk and Grobb city
    entrances, 3 096 units from the northern entrance the zone itself names. With one label, the
    unnamed end is the point of that outline nearest the named doorway, so it stays pinned to a
    real exit rather than to bounding-box geometry.

    The edge's stored `at` is not the doorway either: it holds the drawn connector's endpoints,
    which on Innothule/South Ro sit 16 553 apart - further than the centroids themselves - and
    land on neither outline.

    transformed=False is what ships. Rotation and scale are about the centroid so they cannot
    move it, but tx/ty can - so a transformed cost would shift whenever an author drags a zone
    for looks. Untransformed stored geometry is stable. Both are computed so the review report
    can flag where they disagree.
    """
    z1, z2 = zones[k1], zones[k2]
    fwd = (lambda z, p: tpoint(z, p[0], p[1])) if transformed else (lambda z, p: p)
    c1, c2 = fwd(z1, (z1["cx"], z1["cy"])), fwd(z2, (z2["cx"], z2["cy"]))
    e1 = (exits or {}).get((k1, k2))
    e2 = (exits or {}).get((k2, k1))
    if e1 and e2:
        pa, pb = fwd(z1, e1), fwd(z2, e2)
    elif e1:
        pa = fwd(z1, e1)
        pb = nearest_outline_point(z2, pa[0], pa[1], transformed)
    elif e2:
        pb = fwd(z2, e2)
        pa = nearest_outline_point(z1, pb[0], pb[1], transformed)
    else:
        A, B = cost_points(z1, transformed), cost_points(z2, transformed)
        best, pa, pb = float("inf"), c1, c2
        for ax, ay in A:
            for bx, by in B:
                d = (ax - bx) ** 2 + (ay - by) ** 2
                if d < best:
                    best, pa, pb = d, (ax, ay), (bx, by)
    reach = math.hypot(c1[0] - pa[0], c1[1] - pa[1]) \
        + math.hypot(pb[0] - c2[0], pb[1] - c2[1])
    return reach / UNITS_PER_COST

File: src/mapgeom.js
'use strict';

const MapGeom = (() => {
  const COST_SAMPLE = 200;
  const UNITS_PER_COST = 250.0;

  function roundHalfEven(x) {
    const f = Math.floor(x), d = x - f;
    if (d > 0.5) return f + 1;
    if (d < 0.5) return f;
    return f % 2 === 0 ? f : f + 1;
  }

  function round1(x) {
    const value = roundHalfEven(x * 10) / 10;
    return Object.is(value, -0) ? 0 : value;
  }

  function norm(dx, dy) {
    return Math.sqrt(dx * dx + dy * dy);
  }

  function tpoint(z, x, y) {
    const xf = z.xf || {};
    const s = xf.s == null ? 1 : xf.s;
    const rot = xf.rot || 0;
    const dx = (x - z.cx) * s, dy = (y - z.cy) * s;
    const c = Math.cos(rot), si = Math.sin(rot);
    return [z.cx + dx * c - dy * si + (xf.tx || 0),
      z.cy + dx * si + dy * c + (xf.ty || 0)];
  }

  function tinv(z, X, Y) {
    const xf = z.xf || {};
    const s = xf.s == null ? 1 : xf.s;
    const rot = xf.rot || 0;
    const dx = X - z.cx - (xf.tx || 0), dy = Y - z.cy - (xf.ty || 0);
    const c = Math.cos(-rot), si = Math.sin(-rot);
    return [z.cx + (dx * c - dy * si) / s,
      z.cy + (dx * si + dy * c) / s];
  }

  const SPELL = Object.freeze([
    ['forrest', 'forest'], ['excile', 'exile'], ['cablis', 'cabilis'],
    ['toxullia', 'toxxulia'], ['feerott', 'feerrott'], ['aquaduct', 'aqueduct'],
    ['northern', 'north'], ['southern', 'south'], ['eastern', 'east'], ['western', 'west'],
  ]);
  const ZALIAS = Object.freeze({
    butcherblock: 'butcherblock mountains', 'kerra ridge': 'kerra isle',
    'castle of mistmoore': 'castle mistmoore', 'city of guk': 'upper guk',
    'erudin city': 'erudin', 'erudin docks': 'erudin',
    'north ro': 'north desert of ro', 'south ro': 'south desert of ro',
    'permafrost keep': 'permafrost caverns', 'temple of cazic-thule': 'cazic-thule',
    'skyshrine lower level': 'skyshrine', 'ruins of old guk': 'lower guk',
    'ruins of old paineel': 'warrens', 'ruins of sebilis': 'old sebilis',
    'city of thurgadin': 'thurgadin', 'qeynos aqueduct system': 'qeynos catacombs',
    'liberated citadel of runnyeye': 'runnyeye citadel',
    'valley of king xorbb': 'gorge of king xorbb',
  });
  const DISCOVERY_EXCLUDE = new Set([
    'arcstone', 'arginhiz', 'barren', 'bloodfalls', 'breedinggrounds', 'brellsrest',
    'broodlands', 'commonlands', 'corathus', 'crystalshard', 'delvea', 'dragonscale',
    'eastwastesshard', 'ethernere', 'feerrott2', 'freeportacademy', 'freeportcityhall',
    'freeporthall', 'freeportmilitia', 'freeportsewers', 'freeporttheater', 'freeportwest',
    'gorowyn', 'growthplane', 'gunthak', 'highpasshold', 'jaggedpine', 'kaelshard',
    'korshaext', 'lopingplains', 'mischiefplane', 'mistythicket', 'moors', 'neriakd',
    'oceangreenhills', 'oceanoftears', 'scorchedwoods', 'soldungc', 'steamfontmts',
    'takishruins', 'toxxulia', 'xorbb',
  ]);
  const DISCOVERED_ZONE_COLOR = '#8f78d4';
  const LINK_OVERRIDE = Object.freeze({
    'kithicor|to_The_Commonlands': 'commons',
    'befallen|to_The_Commonlands': 'commons',
  });

  function znorm(value) {
    let s = value.toLowerCase().replace(/`/g, "'").replace(/_/g, ' ');
    s = s.replace(/^\s*(to|from)\s+/, '');
    s = s.replace(/\(.*?\)/g, '');
    s = s.replace(/:.*$/, '');
    s = s.replace(/\bone[- ]way\b/g, '').replace(/&/g, ' ').replace(/ - /g, ' ');
    for (const [a, b] of SPELL) s = s.split(a).join(b);
    s = s.split('plains of karana').join('karana');
    s = s.replace(/^(the|clan)\s+/, '');
    return s.replace(/\s+/g, ' ').replace(/^[ -]+|[ -]+$/g, '');
  }

  function zidxFrom(entries) {
    const idx = Object.create(null);
    function put(raw, value) {
      const key = raw.trim();
      if (key && !Object.prototype.hasOwnProperty.call(idx, key)) idx[key] = value;
    }
    for (const [cont, key, name] of entries) {
      const lower = name.toLowerCase();
      put(znorm(name), { cont, key });
      const directional = lower.match(/\((north|south|east|west)\)/);
      if (directional) {
        const base = znorm(lower.replace(/\s*\((north|south|east|west)\)/, ''));
        put(`${directional[1]} ${base}`, { cont, key });
        put(`${base} ${directional[1]}`, { cont, key });
      }
      const neriak = lower.match(/neriak \((.*)\)/);
      if (neriak) put(`neriak ${neriak[1]}`, { cont, key });
    }
    return idx;
  }

  function resolveZone(idx, label) {
    const normalized = znorm(label);
    if (Object.prototype.hasOwnProperty.call(idx, normalized)) return idx[normalized];
    const alias = ZALIAS[normalized];
    return alias && Object.prototype.hasOwnProperty.call(idx, alias) ? idx[alias] : null;
  }

  function transitionTargets(idx, zoneKey, full) {
    if (!/^(to|from)_/i.test(full)) return [];
    const override = LINK_OVERRIDE[`${zoneKey}|${full}`];
    if (override) return [override];
    const amp = full.indexOf('&');
    const pieces = amp < 0 ? [full] : [full.slice(0, amp), full.slice(amp + 1)];
    const out = [];
    for (const piece of pieces) {
      const target = resolveZone(idx, piece);
      if (target) out.push(target.key);
    }
    return out;
  }

  function discoverySeriesStem(key) {
    key = key.toLowerCase();
    return key.length >= 4 && /^[a-j0-9]$/.test(key.slice(-1)) ? key.slice(0, -1) : null;
  }

  function discoveryDerivedParent(key, roster) {
    key = key.toLowerCase();
    const matches = [];
    for (let parent of roster) {
      parent = parent.toLowerCase();
      const tail = key.startsWith(parent) ? key.slice(parent.length) : null;
      if ((tail !== null && /^(b|c|two|twoa|twob)$/.test(tail)) ||
          key === `old${parent}` || key === `${parent}_original`) matches.push(parent);
    }
    if (!matches.length) return null;
    return matches.reduce((best, value) => value.length > best.length ? value : best);
  }

  function discoveryDisplayName(label) {
    return label.replace(/^(to|from)_/i, '').replace(/\(.*?\)/g, '')
      .replace(/_/g, ' ').replace(/\s+/g, ' ').trim();
  }

  function detailOffset(zone, detail) {
    if (!zone.segs.length || !detail.segs.length) return null;
    const ox = zone.segs[0][0] - detail.segs[0][0];
    const oy = zone.segs[0][1] - detail.segs[0][1];
    const middle = Math.floor(Math.min(zone.segs.length, detail.segs.length) / 2);
    const a = zone.segs[middle], b = detail.segs[middle];
    if (Math.abs((a[0] - b[0]) - ox) > 1.5 || Math.abs((a[1] - b[1]) - oy) > 1.5) return null;
    return [ox, oy];
  }

  function pairKey(from, to) { return `${from}\0${to}`; }

  function exitPointsFrom(zoneKey, zone, detail, idx) {
    const offset = detailOffset(zone, detail);
    const out = new Map();
    if (!offset) return out;
    for (const label of detail.labels) {
      for (const target of transitionTargets(idx, zoneKey, label[4])) {
        const key = pairKey(zoneKey, target);
        if (!out.has(key)) out.set(key, [label[0] + offset[0], label[1] + offset[1]]);
      }
    }
    return out;
  }

  function nearestOutlinePoint(zone, px, py, transformed) {
    let best = Infinity, point = [zone.cx, zone.cy];
    for (const seg of zone.segs) {
      for (const candidate of [[seg[0], seg[1]], [seg[2], seg[3]]]) {
        const q = transformed ? tpoint(zone, candidate[0], candidate[1]) : candidate;
        const distance = (q[0] - px) ** 2 + (q[1] - py) ** 2;
        if (distance < best) { best = distance; point = q; }
      }
    }
    return point;
  }

  function costPoints(zone, transformed) {
    const key = transformed ? '_cpts_t' : '_cpts';
    if (!Object.prototype.hasOwnProperty.call(zone, key)) {
      let points = [];
      for (const seg of zone.segs) points.push([seg[0], seg[1]], [seg[2], seg[3]]);
      if (points.length > COST_SAMPLE) {
        const step = points.length / COST_SAMPLE;
        points = Array.from({ length: COST_SAMPLE }, (_, i) => points[Math.trunc(i * step)]);
      }
      zone[key] = transformed ? points.map(p => tpoint(zone, p[0], p[1])) : points;
    }
    return zone[key];
  }

  function costBetween(zones, key1, key2, transformed, exits = null) {
    const z1 = zones[key1], z2 = zones[key2];
    const fwd = transformed ? (z, p) => tpoint(z, p[0], p[1]) : (_z, p) => p;
    const c1 = fwd(z1, [z1.cx, z1.cy]), c2 = fwd(z2, [z2.cx, z2.cy]);
    const e1 = exits && exits.get(pairKey(key1, key2));
    const e2 = exits && exits.get(pairKey(key2, key1));
    let pa, pb;
    if (e1 && e2) {
      pa = fwd(z1, e1); pb = fwd(z2, e2);
    } else if (e1) {
      pa = fwd(z1, e1); pb = nearestOutlinePoint(z2, pa[0], pa[1], transformed);
    } else if (e2) {
      pb = fwd(z2, e2); pa = nearestOutlinePoint(z1, pb[0], pb[1], transformed);
    } else {
      const left = costPoints(z1, transformed), right = costPoints(z2, transformed);
      let best = Infinity; pa = c1; pb = c2;
      for (const a of left) for (const b of right) {
        const distance = (a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2;
        if (distance < best) { best = distance; pa = a; pb = b; }
      }
    }
    return (norm(c1[0] - pa[0], c1[1] - pa[1]) +
      norm(pb[0] - c2[0], pb[1] - c2[1])) / UNITS_PER_COST;
  }

  return Object.freeze({
    COST_SAMPLE, UNITS_PER_COST, ZALIAS, DISCOVERY_EXCLUDE, DISCOVERED_ZONE_COLOR,
    LINK_OVERRIDE, roundHalfEven, round1, norm, tpoint, tinv, znorm, zidxFrom,
    resolveZone, transitionTargets, discoverySeriesStem, discoveryDerivedParent,
    discoveryDisplayName, detailOffset, exitPointsFrom, nearestOutlinePoint, costPoints,
    costBetween,
  });
})();

if (typeof module !== 'undefined' && module.exports) module.exports = MapGeom;
